dreadnode/agent/tools/fs.py

```diff
@@ -116,23 +133,34 @@ def _relative(self, path: "UPath") -> str:
         return full_path[len(base_path) :]
 
     @tool_method(variants=["read", "write"], catch=True)
-    def read_file(
+    async def read_file(
         self,
         path: t.Annotated[str, "Path to the file to read"],
-    ) -> rg.ContentImageUrl | str:
-        """Read a file and return its contents."""
+    ) -> rg.ContentImageUrl | str | bytes:
+        """
+        Read a file and return its contents.
+
+        Returns:
+            - str: The file contents decoded as UTF-8 if possible.
+            - rg.ContentImageUrl: If the file is non-text and multi_modal is True.
+            - bytes: If the file cannot be decoded as UTF-8 and multi_modal is False.
+
+        Note:
+            Callers should be prepared to handle raw bytes if the file is not valid UTF-8 and multi_modal is False.
+        """
         _path = self._resolve(path)
-        content = _path.read_bytes()
+        async with aiofiles.open(_path, "rb") as f:
+            content = await f.read()
 
         try:
             return content.decode("utf-8")
-        except UnicodeDecodeError as e:
+        except UnicodeDecodeError:
             if self.multi_modal:
                 return rg.ContentImageUrl.from_file(path)
-            raise ValueError("File is not a valid text file.") from e
+            return content
```

Comment:
Are we safe to return bytes here? I think rigging would just call `str(bytes)` on it when building the message, which might be fine. Just double checking.